File: flask-backend/app/services/lead_service.py
"""
services/lead_service.py — All lead business logic.

Routes are HTTP adapters; this is where the actual work happens.
Testable without Flask context (just needs db.session and models).
"""
from ..extensions import db
from ..models.lead import Lead, LEAD_STATUSES
from ..models.user import User
from ..models.note import Note
from .activity_service import ActivityService


class LeadService:

    @staticmethod
    def list_leads(user_id: int, page: int = 1, limit: int = 20,
                   status: str = None, assigned_to: int = None) -> dict:
        user = db.session.get(User, user_id)
        query = Lead.query

        # Members see ONLY their assigned leads — enforced in service, not UI
        if user.role == "member":
            query = query.filter(Lead.assigned_to_id == user_id)
        elif assigned_to:
            query = query.filter(Lead.assigned_to_id == assigned_to)

        if status:
            if status not in LEAD_STATUSES:
                raise ValueError(f"Invalid status '{status}'")
            query = query.filter(Lead.status == status)

        query = query.order_by(Lead.created_at.desc())
        paginated = query.paginate(page=page, per_page=limit, error_out=False)

        return {
            "data": [l.to_dict() for l in paginated.items],
            "pagination": {
                "total": paginated.total,
                "page": paginated.page,
                "limit": paginated.per_page,
                "pages": paginated.pages,
            },
        }

    @staticmethod
    def get_by_id(lead_id: int, requesting_user_id: int) -> Lead:
        lead = db.session.get(Lead, lead_id)
        if not lead:
            raise LookupError("Lead not found")

        user = db.session.get(User, requesting_user_id)
        if user.role == "member" and lead.assigned_to_id != requesting_user_id:
            raise PermissionError("Access denied to this lead")

        return lead

    @staticmethod
    def create(data: dict, created_by: int = None) -> dict:
        lead = Lead(
            name=data["name"],
            email=data["email"],
            phone=data.get("phone"),
            company=data.get("company"),
            source=data.get("source", "Website"),
            message=data.get("message"),
            created_by_id=created_by,
            status="New",
        )
        db.session.add(lead)
        db.session.flush()  # flush to get lead.id before activity log

        ActivityService.log(
            lead_id=lead.id,
            action="Lead created",
            actor_id=created_by,
            extra_data={"source": lead.source},
        )
        db.session.commit()
        return lead.to_dict()

    @staticmethod
    def update(lead_id: int, data: dict, requesting_user_id: int) -> dict:
        user = db.session.get(User, requesting_user_id)
        lead = db.session.get(Lead, lead_id)
        if not lead:
            raise LookupError("Lead not found")

        if user.role == "member" and lead.assigned_to_id != requesting_user_id:
            raise PermissionError("Access denied")

        changes = []
        allowed = ["name", "email", "phone", "company", "source", "message", "status"]

        for field in allowed:
            if field in data and data[field] != getattr(lead, field):
                old_val = getattr(lead, field)
                setattr(lead, field, data[field])
                changes.append(f"{field}: '{old_val}' → '{data[field]}'")

        if changes:
            ActivityService.log(
                lead_id=lead.id,
                action=f"Lead updated: {'; '.join(changes)}",
                actor_id=requesting_user_id,
            )

        db.session.commit()
        return lead.to_dict()

    @staticmethod
    def assign(lead_id: int, assignee_id: int, actor_id: int) -> dict:
        lead = db.session.get(Lead, lead_id)
        if not lead:
            raise LookupError("Lead not found")

        assignee = db.session.get(User, assignee_id)
        if not assignee:
            raise LookupError("Assignee user not found")

        old_assignee = lead.assignee.name if lead.assignee else "Unassigned"
        lead.assigned_to_id = assignee_id

        ActivityService.log(
            lead_id=lead.id,
            action=f"Lead assigned to {assignee.name} (was: {old_assignee})",
            actor_id=actor_id,
            extra_data={"assignee_id": assignee_id},
        )
        db.session.commit()
        return lead.to_dict()

    @staticmethod
    def delete(lead_id: int) -> None:
        lead = db.session.get(Lead, lead_id)
        if not lead:
            raise LookupError("Lead not found")
        db.session.delete(lead)
        db.session.commit()

    @staticmethod
    def add_note(lead_id: int, body: str, author_id: int) -> dict:
        lead = db.session.get(Lead, lead_id)
        if not lead:
            raise LookupError("Lead not found")

        user = db.session.get(User, author_id)
        note = Note(lead_id=lead_id, author_id=author_id, body=body)
        db.session.add(note)
        db.session.flush()

        ActivityService.log(
            lead_id=lead_id,
            action=f"Note added by {user.name}",
            actor_id=author_id,
        )
        db.session.commit()
        return note.to_dict()

    @staticmethod
    def get_notes(lead_id: int) -> list:
        notes = Note.query.filter_by(lead_id=lead_id).order_by(Note.created_at.desc()).all()
        return [n.to_dict() for n in notes]

File: backend/services/leadService.js
const Lead = require('../models/Lead');
const Note = require('../models/Note');
const Activity = require('../models/Activity');
const User = require('../models/User');
const { logActivity } = require('../utils/activityLogger');
const { buildLeadCSV } = require('../utils/csvExport');

/**
 * Build query filters for lead listing
 */
const buildLeadQuery = ({ status, assignedTo, company, search }, requestingUser) => {
  const query = {};

  // Members can only see their assigned leads
  if (requestingUser.role === 'Member') {
    query.assignedTo = requestingUser._id;
  } else {
    if (assignedTo) query.assignedTo = assignedTo;
  }

  if (status) query.status = status;
  if (company) query.company = new RegExp(company, 'i');
  if (search) {
    query.$or = [
      { name: new RegExp(search, 'i') },
      { email: new RegExp(search, 'i') },
      { company: new RegExp(search, 'i') },
    ];
  }

  return query;
};

/**
 * Get paginated leads
 */
const getLeads = async (filters, pagination, requestingUser) => {
  const { page = 1, limit = 10, sort = '-createdAt' } = pagination;
  const skip = (page - 1) * limit;
  const query = buildLeadQuery(filters, requestingUser);

  const [leads, total] = await Promise.all([
    Lead.find(query)
      .populate('assignedTo', 'name email')
      .populate('createdBy', 'name email')
      .sort(sort)
      .skip(skip)
      .limit(Number(limit)),
    Lead.countDocuments(query),
  ]);

  return {
    leads,
    pagination: {
      total,
      page: Number(page),
      limit: Number(limit),
      pages: Math.ceil(total / limit),
    },
  };
};

/**
 * Get a single lead by ID
 */
const getLeadById = async (id, requestingUser) => {
  const lead = await Lead.findById(id)
    .populate('assignedTo', 'name email role')
    .populate('createdBy', 'name email');

  if (!lead) {
    const err = new Error('Lead not found.');
    err.statusCode = 404;
    throw err;
  }

  // Members can only access their assigned leads
  if (
    requestingUser.role === 'Member' &&
    lead.assignedTo?._id?.toString() !== requestingUser._id.toString()
  ) {
    const err = new Error('Access denied to this lead.');
    err.statusCode = 403;
    throw err;
  }

  return lead;
};

/**
 * Create a new lead (public or authenticated)
 */
const createLead = async (data, createdBy = null) => {
  const lead = await Lead.create({ ...data, createdBy, source: createdBy ? 'Manual' : 'Website' });

  await logActivity(
    lead._id,
    'Lead created',
    createdBy,
    { status: lead.status }
  );

  return lead;
};

/**
 * Update lead fields
 */
const updateLead = async (id, updates, requestingUser) => {
  const lead = await Lead.findById(id);
  if (!lead) {
    const err = new Error('Lead not found.');
    err.statusCode = 404;
    throw err;
  }

  const allowedFields = ['name', 'email', 'phone', 'company', 'message', 'status'];
  const changes = [];

  allowedFields.forEach((field) => {
    if (updates[field] !== undefined && updates[field] !== lead[field]) {
      changes.push(`${field} changed from "${lead[field]}" to "${updates[field]}"`);
      lead[field] = updates[field];
    }
  });

  await lead.save();
  await lead.populate('assignedTo', 'name email');

  if (changes.length > 0) {
    await logActivity(lead._id, `Lead updated: ${changes.join('; ')}`, requestingUser._id);
  }

  if (updates.status && updates.status !== lead.status) {
    await logActivity(
      lead._id,
      `Status changed to "${updates.status}"`,
      requestingUser._id,
      { newStatus: updates.status }
    );
  }

  return lead;
};

/**
 * Delete a lead and all associated data
 */
const deleteLead = async (id) => {
  const lead = await Lead.findByIdAndDelete(id);
  if (!lead) {
    const err = new Error('Lead not found.');
    err.statusCode = 404;
    throw err;
  }

  // Cascade delete notes and activities
  await Promise.all([
    Note.deleteMany({ leadId: id }),
    Activity.deleteMany({ leadId: id }),
  ]);

  return lead;
};

/**
 * Assign a lead to a user
 */
const assignLead = async (leadId, userId, requestingUser) => {
  const [lead, assignee] = await Promise.all([
    Lead.findById(leadId),
    User.findById(userId).select('name email role'),
  ]);

  if (!lead) {
    const err = new Error('Lead not found.');
    err.statusCode = 404;
    throw err;
  }

  if (!assignee) {
    const err = new Error('User not found.');
    err.statusCode = 404;
    throw err;
  }

  lead.assignedTo = userId;
  await lead.save();

  await logActivity(
    lead._id,
    `Lead assigned to ${assignee.name}`,
    requestingUser._id,
    { assignedTo: userId }
  );

  await lead.populate('assignedTo', 'name email');
  return lead;
};

/**
 * Add a note to a lead
 */
const addNote = async (leadId, text, requestingUser) => {
  const lead = await Lead.findById(leadId);
  if (!lead) {
    const err = new Error('Lead not found.');
    err.statusCode = 404;
    throw err;
  }

  const note = await Note.create({
    leadId,
    userId: requestingUser._id,
    text,
  });

  await note.populate('userId', 'name email');
  await logActivity(leadId, `Note added by ${requestingUser.name}`, requestingUser._id);

  return note;
};

/**
 * Get notes for a lead
 */
const getNotes = async (leadId) => {
  return Note.find({ leadId })
    .populate('userId', 'name email')
    .sort({ createdAt: -1 });
};

/**
 * Get activity timeline for a lead
 */
const getActivities = async (leadId) => {
  return Activity.find({ leadId })
    .populate('performedBy', 'name email')
    .sort({ createdAt: -1 });
};

/**
 * Get admin dashboard analytics
 */
const getDashboardStats = async () => {
  const [statusCounts, totalLeads, recentActivities] = await Promise.all([
    Lead.aggregate([
      { $group: { _id: '$status', count: { $sum: 1 } } },
    ]),
    Lead.countDocuments(),
    Activity.find()
      .populate('performedBy', 'name')
      .populate('leadId', 'name email')
      .sort({ createdAt: -1 })
      .limit(10),
  ]);

  const stats = {
    total: totalLeads,
    new: 0,
    contacted: 0,
    qualified: 0,
    proposalSent: 0,
    won: 0,
    lost: 0,
    conversionRate: 0,
  };

  const keyMap = {
    New: 'new',
    Contacted: 'contacted',
    Qualified: 'qualified',
    'Proposal Sent': 'proposalSent',
    Won: 'won',
    Lost: 'lost',
  };

  statusCounts.forEach(({ _id, count }) => {
    if (keyMap[_id]) stats[keyMap[_id]] = count;
  });

  if (totalLeads > 0) {
    stats.conversionRate = Math.round((stats.won / totalLeads) * 100);
  }

  return { stats, recentActivities };
};

/**
 * Export leads to CSV
 */
const exportLeadsCSV = async (filters, requestingUser) => {
  const query = buildLeadQuery(filters, requestingUser);
  const leads = await Lead.find(query).lean();
  return buildLeadCSV(leads);
};

module.exports = {
  getLeads,
  getLeadById,
  createLead,
  updateLead,
  deleteLead,
  assignLead,
  addNote,
  getNotes,
  getActivities,
  getDashboardStats,
  exportLeadsCSV,
};
